src/matrices.rs

```diff
@@ -0,0 +1,580 @@
+            }
+        }
+        Self::new(res)
+    }
+
+    fn zeros((h, w): (usize, usize)) -> Self {
+        let mut res = vec![];
+        for _ in 0..h {
+            res.push(vec![T::zero(); w]);
+        }
+        Self::new(res)
+    }
+
+    fn identity(n: usize) -> Self {
+        Self::filled((n, n), |i, j| if i == j { T::one() } else { T::zero() })
+    }
+
+    fn echelon(mut self) -> anyhow::Result<Aftermath<T>> {
+        const CONTEXT: &str = "Calculations error!";
+
+        if self.data.is_empty() {
+            return Ok(Aftermath::<T> {
+                result: self,
+                steps: Vec::new(),
+            });
+        }
+
+        let rows = self.data.len();
+        let cols = self.data[0].len();
+
+        let mut steps = vec![self.to_latex()];
+        let mut c = 0;
+        let mut i = 0;
+
+        let mut add_step = |step: &str, matrix: &Matrix<T>| {
+            steps.push(format!(r"\xrightarrow{{{}}} {}", step, matrix.to_latex()));
+        };
+
+        fn nice<T: MatrixNumber>(coef: &T) -> Option<i64> {
+            if coef.is_zero() {
+                Some(1000)
+            } else if coef.is_one() {
+                Some(0)
+            } else if (T::zero().checked_sub(coef)?).is_one() {
+                Some(1)
+            } else {
+                Some(2)
+            }
+        }
+
+        fn sub_coeff_to_latex<T: MatrixNumber>(coef: &T) -> Option<String> {
+            if coef.is_one() {
+                Some("- ".to_string())
+            } else if (T::zero().checked_sub(coef)?).is_one() {
+                Some("+ ".to_string())
+            } else if coef.is_positive() {
+                Some(format!("- {}", coef.to_latex()))
+            } else if coef.is_negative() {
+                Some(format!("+ {}", (T::zero().checked_sub(coef)?.to_latex())))
+            } else {
+                unreachable!("Should not be used for zero coefficient!")
+            }
+        }
+
+        while c < cols && i < rows {
+            let mut j = i;
+            for k in i + 1..rows {
+                if nice(&self.data[k][c]).context(CONTEXT)?
+                    < nice(&self.data[j][c]).context(CONTEXT)?
+                {
+                    j = k;
+                }
+            }
+
+            if !self.data[j][c].is_zero() {
+                if i != j {
+                    self.data.swap(i, j);
+                    add_step(
+                        format!(r"w_{{{}}} \leftrightarrow w_{{{}}}", i + 1, j + 1).as_str(),
+                        &self,
+                    );
+                }
+
+                if !self.data[i][c].is_one() {
+                    let d = self.data[i][c].clone();
+                    for j in c..cols {
+                        self.data[i][j] = self.data[i][j].checked_div(&d).context(CONTEXT)?;
+                    }
+
+                    add_step(
+                        format!(r"w_{{{}}} : {}", i + 1, d.to_latex_single()).as_str(),
```

Comment:
za długa jest ta metoda i zbyt skomplikowana, przydałoby się podnieść poziom abstrakcji, podzielić to na kawałki i wprowadzić sensowne nazwy